### PR TITLE
Support IMessageActivity on the Context Send methods. This allows the MessageFactory to easily send.

### PR DESCRIPTION
**Pull request by cleemullins**
Addresses #282

### CODE REVIEW

**Comment by brandonh-msft** [comment]
Why not `IActivity`?? What if I want to send a typing activity?

File: libraries/Microsoft.Bot.Builder.Core/BotContext.cs [hunk 1]
```diff
@@ -74,6 +74,14 @@ public bool Responded
             }
         }
 
+        public async Task SendActivity(params IMessageActivity[] messagesToSend)
```
Comment (brandonh-msft) [inline libraries/Microsoft.Bot.Builder.Core/BotContext.cs]: this should not be `IMessageActivity` - what if the user wants to send a `Typing` activity? A conversation activity? Or a <insert other activity type here> Activity?

File: libraries/Microsoft.Bot.Builder.Core/IBotContext.cs [hunk 2]
```diff
@@ -28,6 +28,8 @@ public interface IBotContext
 
         Task SendActivity(params string[] textRepliesToSend);
         Task SendActivity(params Activity[] activities);
+        Task SendActivity(params IMessageActivity[] activities);
```
Comment (brandonh-msft) [inline libraries/Microsoft.Bot.Builder.Core/IBotContext.cs]: We shouldn't be simply adding overloads for every possible type.

Comment (cleemullins) [reply libraries/Microsoft.Bot.Builder.Core/IBotContext.cs]: Good points Look for the checkin shortly.

**Comment by cleemullins** [comment]
@brandonh-msft , code is updated to the IActivity interface, per your suggestion. The downside is that it means the Event handlers that people use will be using the interfaces, rather than the class, which many seem to find frustrating. On the SendActivity() / UpdateActivity() side though, it does clean things up.

**Comment by brandonh-msft** [comment]
We talked about this a long time ago (`Activity` vs `IActivity`) and you had stated that cleanup was "coming soon ™ so I assume this helps align that work ;)

**Comment by cleemullins** [comment]
I don't think we have resolution yet as to a longer term plan. To a large degree, we're constrained by the back-compat in the lower layers as we've made the deliberate decision not to break that. 

For now, this seems like it'l let developers at least pass the Messages around more easily.

File: libraries/Microsoft.Bot.Builder.Core/BotContext.cs [hunk 3]
```diff
@@ -107,15 +99,20 @@ public bool Responded
                 ApplyConversationReference(a, cr);
             }
 
-            List<Activity> activityList = new List<Activity>(activities);
+            // Convert the IActivities to Activies. 
+            Activity[] activityArray = Array.ConvertAll(activities, (input) => (Activity)input);            
+
+            // Create the list used by the recursive methods. 
+            List<Activity> activityList = new List<Activity>(activityArray);
```
Comment (brandonh-msft) [inline libraries/Microsoft.Bot.Builder.Core/BotContext.cs]: at this point you've now got 3 collections of the same objects; `activities`, `activityArray`, and `activityList`. I'm thinking `activities.Select(a=>(Activity)a).ToList()` would be much more efficient.

File: libraries/Microsoft.Bot.Builder.Core/BotContext.cs [hunk 4]
```diff
@@ -99,15 +99,20 @@ public bool Responded
...
 
             async Task ActuallySendStuff()
             {
                 bool anythingToSend = false;
                 if (activities.Count() > 0)
                     anythingToSend = true;
 
-                await this.Adapter.SendActivity(this, activities);
+                // Send from the list, which may have been manipulated via the event handlers. 
+                await this.Adapter.SendActivity(this, activityList.ToArray());
```
Comment (brandonh-msft) [inline libraries/Microsoft.Bot.Builder.Core/BotContext.cs]: `ToArray()` creates yet another copy of the same items...

**Comment by tomlm** [comment]
>             await SendActivitiesInternal(activityList, _onSendActivities, ActuallySendStuff);

Is this what you are trying to do?
```
// send all activities 
foreach(var activity in activites)
{
       await this.Adapter.SendActivity(this, (Activity)activity);
}
if (!this.Responded & activities.Any())
    this.Responded = true;
```


---
Refers to: libraries/Microsoft.Bot.Builder.Core/BotContext.cs:122 in 3915c08. [](commit_id = 3915c0822b392ff8ac991dc0f545a88b1fa0f77d, deletion_comment = False)

**Pull request merged**